s21/p5/tester.py

```diff
@@ -0,0 +1,509 @@
+                    print(f"ip: {expected_dict['ip']}, expected ip_int {expected_dict['int_ip']} but found {actual_dict['int_ip']}")
+                    points -= unit_points
+                else:
+                    if actual_dict['region'] != expected_dict['region']: # region check
+                        print(f"ip: {expected_dict['ip']}, expected region {expected_dict['region']} but found {actual_dict['region']}")
+                        points -= unit_points
+        except IndexError as e:
+            print(f"missing: {expected[i]['ip']}")
+        except Exception as e:
+            print(type(e), e)
+    
+    # optimize: faster processing on consecutive ones, criteria: < 20% of random access
+    avg_time_inconsecutive = np.mean([x['ms'] for x in actual[:6]])
+    avg_time_consecutive = np.mean([x['ms'] for x in actual[6:]])
+    # print('time', avg_time_inconsecutive, avg_time_consecutive)
+    
+    if (avg_time_inconsecutive / 10) < avg_time_consecutive:
+        print("(-10) Not enough optimized for consecutive ips! Time for consecutive ips can be less than 10% of inconsecutive ones")
+        print(f"average elapsed time for consecutive ips: {avg_time_consecutive}")
+        print(f"average elapsed time for inconsecutive ips: {avg_time_inconsecutive}")
+        points -= 10
+    return max(points, 0)
+
+@test(points=25)
+def sample():
+    points = 25
+    zname = "large.zip"
+    zout = "small_samp.zip"
+    stride = 1000
+    start_time = time.time()
+    try:
+        run("sample", zname, zout, stride) # include stride of test data set
+    except subprocess.TimeoutExpired:
+        print("Execution time for sample should be less than 90 seconds")
+        return 0
+    except Exception as e:
+        print(type(e), e)
+        
+    elapsed_time = time.time() - start_time
+    if elapsed_time > 20:
+        print("(-10) sample command can be optimized further. (<= 20secs), elapsed time:", elapsed_time)
+        print("Make sure you have an O(M+N) implementation")
+        points -= 10
+        
+    
+    # write on actual_json
+    actual_count = 0
+    for i, row in enumerate(zip_csv_iter(zout)):
+        actual_json[f"sample_{i}"] = row
+        actual_count += 1
+    
+    # dump in the middle for convenience
+    with open("actual.json", "w", encoding="utf-8") as f:
+        json.dump(actual_json, f, indent=2)
+    
+    # open expected
+    with open("expected.json") as f:
+        expected_json = json.load(f)
+    
+    expected_count = 0
+    for key in expected_json:
+        if "sample" in key:
+            expected_count += 1
+    
+    unit_points = points / expected_count
+    
+    # compare the number of rows
+    if expected_count != actual_count:
+        print(f"expected {expected_count} rows, found {actual_count} rows.")
+        points -= unit_points * abs(expected_count - actual_count)
+    
+    for i in range(expected_count):
```

Comment:
does anything in this loop actually deduct points?  I also worry about how many prints it will generate if every line is different.

Reply:
* I missed points deduction. Thanks!
* If every line is different, it will generate 175 prints. Are they too many?

Reply:
Now it only prints 5 incorrect cases.